### PR TITLE
[SYCL] Use CUDA symbol for specialization constants

### PR DESCRIPTION
**Pull request by jchlanda**
This patch improves support for CUDA specialization constants. The approach is hybrid, where we make use of spec constant emulation (as already present in oneAPI), which bundles all spec constants together and passes them to the kernel as an additional, implicit argument; as well as special handling in the CUDA pi plugin alongside a custom pass.
The pass creates a global entry that corresponds to the implicit argument and rewrites the kernel to drop that additional implicit argument, furthermore it changes the accesses from implicit argument to the global value. At the runtime side, CUDA pi plugin is responsible for retrieving the symbol (using `cuModuleGetGlobal`) and setting it up (using `cuMemcpyHtoD`).

The patch requires an update to the PI interface, extending the list of arguments of `piextProgramSetSpecializationConstant` with a `pi_kernel`. This is needed, as emulated specialization constants rely on extra kernel argument, for this patch the argument is replaced with a CUDA symbol.

### CODE REVIEW

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:103]:
```suggestion
           "Expected a specialisation constants buffer argument. The function is not a kernel?");
```

?

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:112]:
```suggestion
    for (unsigned I = 0; I < F->arg_size() - 1; ++I) {
      Arguments.push_back(F->getArg(I)->getType());
      ArgumentAttributes.push_back(FAttrs.getParamAttrs(I));
```

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:131]:
```suggestion
      auto *I = dyn_cast_or_null<Instruction>(&*U);
```

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:136]:
```suggestion
        raw_string_ostream Out(Str);
```

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:143]: You should just use [`PointerType::withSamePointeeType`](https://llvm.org/doxygen/classllvm_1_1PointerType.html#ac0f741735997264d3018e30ce2a6b2d3) right away

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:61]: This is a huge method (around 120 lines): what about outlining some "steps" into separate functions? I see like three major steps here: create global variables, change kernel signatures, rewrite kernel argument uses.

File: llvm/tools/sycl-post-link/SpecConstants.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/SpecConstants.cpp:239]: Good catch, we can merge it as a separate `NFC` PR quicker, if you prefer.

**Review by AlexeySachkov — COMMENTED** [review]
`sycl-post-link` part feedback

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:20]: Ha, that's actually a bug, as it would allocate in 64-bit int elements, changing to use 8-bit and `getInt8Ty` as suggested. Done now.

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:57]: Yeap, I've seen llvm generate something along the lines of: `!{null, !"kernel", i32 1}`

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:118]: :) that's the clang tidy warning that I have off, will update

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp:67]: Agreed, will split it to helpers.

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp]: Done

File: llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDANativeSpecConstants.cpp]: Done in: 4f91cb850c55941b086733b7d77b482aa3b5f41c

File: llvm/tools/sycl-post-link/SpecConstants.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/SpecConstants.cpp:239]: Done here: https://github.com/intel/llvm/pull/7992

**Comment by jchlanda** [comment]
Ping, the AMD failures are unrelated, but blocking the CUDA test.

**Comment by jchlanda** [comment]
Ping @intel/llvm-reviewers-cuda @intel/llvm-reviewers-runtime @intel/dpcpp-esimd-reviewers @intel/dpcpp-l0-pi-reviewers @sergey-semenov @AlexeySachkov

File: sycl/include/sycl/detail/pi.h
Comment (smaslov-intel) [inline sycl/include/sycl/detail/pi.h:82]: We are not allowing ABI breaking changes currently, so please add a new version instead.
+@kbenzie, please note this change to PI API and take it into Unified Runtime

File: sycl/include/sycl/detail/pi.h
Comment (smaslov-intel) [inline sycl/include/sycl/detail/pi.h:1333]: Perhaps `kernel` argument would better be positioned after `program`

File: sycl/include/sycl/detail/pi.h
Comment (jchlanda) [reply sycl/include/sycl/detail/pi.h:82]: Do you mean revert `_PI_H_VERSION_MAJOR` back to `12` and keep the change to `_PI_H_VERSION_MINOR` at `24`?

Comment (kbenzie) [reply sycl/include/sycl/detail/pi.h:82]: I've created https://github.com/oneapi-src/unified-runtime/issues/269 to track these changes.

File: sycl/include/sycl/detail/pi.h
Comment (jchlanda) [reply sycl/include/sycl/detail/pi.h]: Agreed, done in: d5f876ef07aca617d1844c59a8deee8eb90b3e40

**Comment by gmlueck** [comment]
This sounds like a large-ish change to the design for specialization constants on CUDA.  Could you add a new section to the "SYCL2020-SpecializationConstants.md" design document explaining how it works on CUDA?

File: sycl/include/sycl/detail/pi.h
Comment (smaslov-intel) [reply sycl/include/sycl/detail/pi.h:82]: > Do you mean revert `_PI_H_VERSION_MAJOR` back to `12` and keep the change to `_PI_H_VERSION_MINOR` at `24`?

Yes, but more importantly re-do changes so that existing interfaces are not changed (and add new ones)

Comment (jchlanda) [reply sycl/include/sycl/detail/pi.h:82]: I'm not sure why we would need the new entry, I thought it was OK to introduce PI API changes (without waiting for the breaking window)?

**Comment by jchlanda** [comment]
Ping, the AMD failures are unrelated, but blocking the CUDA test.

> This sounds like a large-ish change to the design for specialization constants on CUDA. Could you add a new section to the "SYCL2020-SpecializationConstants.md" design document explaining how it works on CUDA?

Agreed, done in: 95ba4822e6904e704b4b5426e0dfbb43f0d1bac9

File: sycl/source/detail/program_impl.cpp
Comment (gmlueck) [inline sycl/source/detail/program_impl.cpp:558]: Why is it OK to pass null here and in the other calls to this PI API?  Are these code paths not used by CUDA?

File: sycl/include/sycl/detail/pi.h
Comment (smaslov-intel) [reply sycl/include/sycl/detail/pi.h:82]: Because PI API crosses libraries boundaries (plugins <-> SYCL RT) and should be stable.
We've been following https://github.com/intel/llvm/blob/sycl/sycl/doc/developer/ABIPolicyGuide.md
I see we need to define there what the next ABI breaking windows is: @stdale-intel ?

File: sycl/source/detail/program_impl.cpp
Comment (jchlanda) [reply sycl/source/detail/program_impl.cpp:558]: That's right (apart from the implementation of `piextProgramSetSpecializationConstant` that each PI backend has), the uses of set spec constants can be broadly split in two different cases:
* `setKernelParamAndLaunch`,
* different flavour of build (`program_impl::build`, `ProgramManager::getBuiltPIProgram`).

Because of the emulation of spec constants and the fact we don't do AOT compilation for Cuda, it's safe to `null`-default the calls in the build and friends (that path will never be exercised) and only support the hook at setting kernel params.

Comment (gmlueck) [reply sycl/source/detail/program_impl.cpp:558]: I'll defer to the @intel/llvm-reviewers-runtime, but I wonder if there should instead be some new PI API for the `setKernelParamAndLaunch` case.  At present, the `setKernelParamAndLaunch` case is called only for the CUDA backend and it seems like it has a different meaning than the other calls to `piextProgramSetSpecializationConstant `.  For example, the `setKernelParamAndLaunch` is setting the "blob" of all spec constants, not an individual spec constant.

If there were two separate APIs, then you would only need a "kernel" parameter for the `setKernelParamAndLaunch` case.  This makes sense because in the `ProgramManager::getBuiltPIProgram` case, there is no relevant kernel.

Comment (jchlanda) [reply sycl/source/detail/program_impl.cpp:558]: I agree, this approach overloads the meaning of `piextProgramSetSpecializationConstant` quite a bit. I would like to propose that we allow the change, with the time ticking on the PI plugins it feels like doing an overhaul now might not be worth it. UR team is aware of the change and tracks it in their issue tracker (https://github.com/oneapi-src/unified-runtime/issues/269).
Alternatively we could tackle it in a follow up PR.

File: sycl/include/sycl/detail/pi.h
Comment (jchlanda) [reply sycl/include/sycl/detail/pi.h:82]: We've discussed this in the call yesterday and came to a conclusion that PI's API changes do not need to wait for the ABI break window, @smaslov-intel are you OK with bumping the major/minor versions and updating the signature of `piextProgramSetSpecializationConstant`?

**Comment by bader** [comment]
> /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/spirv.hpp:594:17: error: call to function 'SubgroupShuffle' that is neither visible in the template definition nor found by argument-dependent lookup

https://github.com/intel/llvm/actions/runs/4244942104/jobs/7380586829

**Comment by bader** [comment]
> > /__w/llvm/llvm/toolchain/bin/../include/sycl/detail/spirv.hpp:594:17: error: call to function 'SubgroupShuffle' that is neither visible in the template definition nor found by argument-dependent lookup
> 
> https://github.com/intel/llvm/actions/runs/4244942104/jobs/7380586829

This seems to be a regression introduced by some other change. I see the same problem in https://github.com/intel/llvm/pull/8276 pre-commit testing results.

**Comment by jchlanda** [comment]
@gmlueck @bader @smaslov-intel @AlexeySachkov is there anything else I should be addressing in this PR (gentle ping)?

File: sycl/source/detail/program_impl.cpp
Comment (gmlueck) [reply sycl/source/detail/program_impl.cpp:558]: Responding because of your "ping" ... it sounds like we both agree that this PR is a little hacky because it overloads the meaning of `piextProgramSetSpecializationConstant` and it would be better to add a new PI API.  It seems like the only question is whether it makes sense to spend the effort to design the new API now vs. waiting for the transition to the UR.

My opinion is that it makes sense to spend the time now because that effort won't be wasted.  If we come up with a nice PI API now, we will reuse it for the UR.

As I said, though, I defer to the development team reviewers on this.  If the team thinks it makes more sense to delay cleanup to the time when we add support for the UR, I have no concerns.

Comment (jchlanda) [reply sycl/source/detail/program_impl.cpp:558]: We've discussed this with the UR team and they are happy to drive the change, the new API entry has already been worked on, please see: https://github.com/oneapi-src/unified-runtime/pull/306

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:197]: I wonder, why not `SmallVector`?

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:153]: Do we expect `&*U` to result in `nullptr`? Because if not, there is no reason to use `dyn_cast_or_null` and we should just use regular `dyn_cast`.

Moreover, if there is no runtime check for `I != nullptr`, then a regular `cast` should be used

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:130]:
```suggestion
    for (unsigned I = 1; I < Node->getNumOperands(); ++I) {
      MDNode *SC = cast<MDNode>(Node->getOperand(I));
      assert(SC->getNumOperands() >= 4 && "Invalid node.");
```

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (AlexeySachkov) [inline llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:54]:
```suggestion
  return static_cast<unsigned>(mdconst::extract<ConstantInt>(
       Node->getOperand(OpNumber).get())->getZExtValue());
```

[Doxygen](https://llvm.org/doxygen/namespacellvm_1_1mdconst.html#ad938857d6c6603847adf3a8cbe403d17)

**Review by AlexeySachkov — APPROVED** [review]
`sycl-post-link` part LGTM. A few (rather minor) comments for the new pass itself.

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:153]: The honest answer is that I don't know, it boils down to how the use list is constructed, and it is not clear from the code, it calls `addUse` (https://github.com/jchlanda/llvm/blob/jakub/cuda_native_symbols/llvm/include/llvm/IR/Value.h#L
505) without validating what the `Use` is.

For what it's worth, a quick grep reveals that the code base has all sorts of approaches:
* `dyn_cast_or_null`: https://github.com/jchlanda/llvm/blob/jakub/cuda_native_symbols/llvm/lib/Transforms/IPO/FunctionSpecialization.cpp#L598 (in the call to `getUserBonus`)
* `dyn_cast`: https://github.com/jchlanda/llvm/blob/jakub/cuda_native_symbols/llvm/lib/Transforms/Utils/SimplifyLibCalls.cpp#L2599 (in the call to `classifyArgUse`)

and plenty of straight up accesses to the casted instance.

I guess, we don't expect it to ever by `null`, but have that sanity check `assert(I && "Expected an instruction.");`, will change to `dyn_cast`.

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:54]: Nice.

File: llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp
Comment (jchlanda) [reply llvm/tools/sycl-post-link/CUDASpecConstantToSymbol.cpp:196]: Good spot, changed.

File: sycl/source/detail/program_impl.cpp
Comment (kbenzie) [reply sycl/source/detail/program_impl.cpp:558]: The UR team are happy to handle the necessary future changes as part of our ongoing effort to port the CUDA pluign to UR. As such, I don't see a need to block this PR based on the design direction that UR will take to enable this functionality.

**Review by steffenlarsen — APPROVED** [review]
Runtime and CUDA implementation looks okay to me, though I share @gmlueck's concerns about the API. It seems like adding a new API would avoid both the overloading of the behavior and the PI ABI break. However, based the latest information in both of the related discussions I don't think it has to be a blocker. @smaslov-intel - Feel free to overrule if you feel strong about it.

**Comment by smaslov-intel** [comment]
> Runtime and CUDA implementation looks okay to me, though I share @gmlueck's concerns about the API. It seems like adding a new API would avoid both the overloading of the behavior and the PI ABI break. However, based the latest information in both of the related discussions I don't think it has to be a blocker. @smaslov-intel - Feel free to overrule if you feel strong about it.

We are in agreement that we do allow PI API/ABI break now, so let's go ahead here. The better interfaces are being added to UR.

**Review by smaslov-intel — APPROVED** [review]

**Comment by jchlanda** [comment]
Switching to draft, as the faith of cuda symbols for spec constants is somehow uncertain.

**Pull request closed without merging**
